OffscreenCanvas WebGL doesn't fire the context lost event.

mattwoodrow opened the pull request:
#### 903910f95dd36fc7cbc660fb96daab4df0372a07
<pre>
OffscreenCanvas WebGL doesn&apos;t fire the context lost event.
<a href="https://bugs.webkit.org/show_bug.cgi?id=250532">https://bugs.webkit.org/show_bug.cgi?id=250532</a>
&lt;rdar://104198422&gt;

Reviewed by Ryosuke Niwa.

Adds support for dispatching the event to OffscreenCanvas too, and enables the WebGL tests that cover this.

* LayoutTests/TestExpectations:
* LayoutTests/platform/mac-wk2/TestExpectations:
* Source/WebCore/html/CanvasBase.h:
* Source/WebCore/html/CustomPaintCanvas.h:
* Source/WebCore/html/HTMLCanvasElement.cpp:
(WebCore::HTMLCanvasElement::queueTaskKeepingObjectAlive):
(WebCore::HTMLCanvasElement::dispatchEvent):
* Source/WebCore/html/HTMLCanvasElement.h:
* Source/WebCore/html/OffscreenCanvas.cpp:
(WebCore::OffscreenCanvas::create):
(WebCore::OffscreenCanvas::OffscreenCanvas):
(WebCore::OffscreenCanvas::transferToImageBitmap):
(WebCore::OffscreenCanvas::queueTaskKeepingObjectAlive):
(WebCore::OffscreenCanvas::dispatchEvent):
* Source/WebCore/html/OffscreenCanvas.h:
* Source/WebCore/html/OffscreenCanvas.idl:
* Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
(WebCore::WebGLRenderingContextBase::scheduleTaskToDispatchContextLostEvent):
(WebCore::WebGLRenderingContextBase::maybeRestoreContext):
* Source/WebCore/html/canvas/WebGLSync.cpp:
(WebCore::WebGLSync::scheduleAllowCacheUpdate):

Canonical link: <a href="https://commits.webkit.org/260924@main">https://commits.webkit.org/260924@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/20462a309cbbeeb86c5b5625b541fd41d2906e59

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/109950 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/19056 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/42639 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/1399 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/119005 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/113901 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/20518 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/10244 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/102212 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/115699 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/15275 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/98490 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/43493 "Found 2 new test failures: webgl/2.0.y/conformance/offscreencanvas/offscreencanvas-transfer-image-bitmap.html, webgl/2.0.y/conformance2/offscreencanvas/offscreencanvas-transfer-image-bitmap.html (failure)") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/97237 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/30146 "Found 1 new test failure: imported/w3c/web-platform-tests/preload/onerror-event.html (failure)") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/85317 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/11758 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/31487 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/12383 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/8441 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/17751 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/51097 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7575 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/14176 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/html/OffscreenCanvas.cpp:
```diff
@@ -576,6 +577,20 @@ void OffscreenCanvas::reset()
+void OffscreenCanvas::queueTaskKeepingObjectAlive(TaskSource source, Function<void()>&& task)
+{
+    UNUSED_PARAM(source);
+    scriptExecutionContext()->postTask([protectedThis = Ref { *this }, task = WTFMove(task)] (ScriptExecutionContext&) mutable {
```
This ain't right. You're not keeping the wrapper alive.

rniwa replied:
I don't think we want to add this function at all. Use the one on ActiveDOMObject instead.

rniwa replied:
Please refer to https://github.com/WebKit/WebKit/blob/main/Introduction.md#js-wrapper-lifecycle-management

rniwa replied:
Also relevant: https://github.com/WebKit/WebKit/blob/main/Introduction.md#creating-a-pending-activity

mattwoodrow replied:
Thanks for the review!

I'll add ActiveDOMObject to OffscreenCanvas, since it seems that's what's needed.

rniwa replied:
Yeah, that would make most sense to me.

rniwa commented on Source/WebCore/html/CanvasBase.h:
```diff
@@ -130,6 +132,9 @@ class CanvasBase {
     WEBCORE_EXPORT static void setMaxCanvasAreaForTesting(std::optional<size_t>);
 
+    virtual void queueTaskKeepingObjectAlive(TaskSource, Function<void()>&&) { };
+    virtual void dispatchEvent(Event&) { }
```
These are better left as pure virtual functions.

rniwa commented on Source/WebCore/html/OffscreenCanvas.h:
```diff
@@ -149,6 +150,11 @@ class OffscreenCanvas final : public RefCounted<OffscreenCanvas>, public CanvasB
 
     void commitToPlaceholderCanvas();
 
+
```
An extra new line here.

mattwoodrow committed:
OffscreenCanvas WebGL doesn't fire the context lost event.
https://bugs.webkit.org/show_bug.cgi?id=250532
<rdar://104198422>

Reviewed by Ryosuke Niwa.

Adds support for dispatching the event to OffscreenCanvas too, and enables the WebGL tests that cover this.

* LayoutTests/TestExpectations:
* LayoutTests/platform/mac-wk2/TestExpectations:
* Source/WebCore/html/CanvasBase.h:
* Source/WebCore/html/CustomPaintCanvas.h:
* Source/WebCore/html/HTMLCanvasElement.cpp:
(WebCore::HTMLCanvasElement::queueTaskKeepingObjectAlive):
(WebCore::HTMLCanvasElement::dispatchEvent):
* Source/WebCore/html/HTMLCanvasElement.h:
* Source/WebCore/html/OffscreenCanvas.cpp:
(WebCore::OffscreenCanvas::create):
(WebCore::OffscreenCanvas::OffscreenCanvas):
(WebCore::OffscreenCanvas::transferToImageBitmap):
(WebCore::OffscreenCanvas::queueTaskKeepingObjectAlive):
(WebCore::OffscreenCanvas::dispatchEvent):
* Source/WebCore/html/OffscreenCanvas.h:
* Source/WebCore/html/OffscreenCanvas.idl:
* Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
(WebCore::WebGLRenderingContextBase::scheduleTaskToDispatchContextLostEvent):
(WebCore::WebGLRenderingContextBase::maybeRestoreContext):
* Source/WebCore/html/canvas/WebGLSync.cpp:
(WebCore::WebGLSync::scheduleAllowCacheUpdate):

Canonical link: https://commits.webkit.org/260924@main
